File: distribution.go
package main

import (
	"fmt"
	"log"
	"math"

	"gocv.io/x/gocv"
	"gonum.org/v1/gonum/mat"
	"gonum.org/v1/gonum/stat"
)

// detectAndCompute detects keypoints and computes descriptors using the provided detector
func detectAndCompute(image gocv.Mat, detector gocv.SIFT) ([]gocv.KeyPoint, gocv.Mat) {
	keypoints, descriptors := detector.DetectAndCompute(image, gocv.NewMat())
	return keypoints, descriptors
}

// Function to calculate spatial distribution score
func calculateSpatialDistribution(keypoints []gocv.KeyPoint, imageSize []int) float64 {
	if len(keypoints) < 2 {
		return 0 // Return 0 if there are not enough keypoints
	}

	xCoords := make([]float64, len(keypoints))
	yCoords := make([]float64, len(keypoints))
	for i, kp := range keypoints {
		xCoords[i] = float64(kp.X)
		yCoords[i] = float64(kp.Y)
	}
	numBins := 10
	xHist := make([]float64, numBins)
	yHist := make([]float64, numBins)
	xBins := make([]float64, numBins+1)
	yBins := make([]float64, numBins+1)
	for i := 0; i <= numBins; i++ {
		xBins[i] = float64(i) * float64(imageSize[1]) / float64(numBins)
		yBins[i] = float64(i) * float64(imageSize[0]) / float64(numBins)
	}

	stat.Histogram(xHist, xCoords, nil, xBins)
	stat.Histogram(yHist, yCoords, nil, yBins)
	spatialDistributionScore := stat.Entropy(xHist) + stat.Entropy(yHist)
	return spatialDistributionScore
}

// Function to calculate descriptor uniqueness score
func calculateUniquenessScore(descriptors gocv.Mat) float64 {
	rows := descriptors.Rows()
	if rows == 0 {
		return 0 // Return 0 if there are no descriptors
	}

	data := make([]float64, rows*rows)
	for i := 0; i < rows; i++ {
		for j := 0; j < rows; j++ {
			if i != j {
				rowI := descriptors.RowRange(i, i+1)
				rowJ := descriptors.RowRange(j, j+1)
				diffMat := gocv.NewMat()
				gocv.Subtract(rowI, rowJ, &diffMat)
				distance := gocv.Norm(diffMat, gocv.NormL2)
				diffMat.Close()
				rowI.Close()
				rowJ.Close()
				data[i*rows+j] = distance
			} else {
				data[i*rows+j] = math.MaxFloat64
			}
		}
	}
	distances := mat.NewDense(rows, rows, data)
	minDistances := make([]float64, rows)
	for i := 0; i < rows; i++ {
		minDistances[i] = mat.Min(distances.RowView(i))
	}
	uniquenessScore := stat.Mean(minDistances, nil)
	return uniquenessScore
}

// Function to calculate coverage score
func calculateCoverage(keypoints []gocv.KeyPoint, imageSize []int) float64 {
	height, width := imageSize[0], imageSize[1]
	gridSize := 10
	grid := make([][]bool, gridSize)
	for i := range grid {
		grid[i] = make([]bool, gridSize)
	}
	for _, kp := range keypoints {
		x, y := int(kp.X*float64(gridSize)/float64(width)), int(kp.Y*float64(gridSize)/float64(height))
		if x >= 0 && x < gridSize && y >= 0 && y < gridSize {
			grid[y][x] = true
		}
	}
	coverageScore := 0
	for _, row := range grid {
		for _, covered := range row {
			if covered {
				coverageScore++
			}
		}
	}
	return float64(coverageScore) / float64(gridSize*gridSize)
}

// Function to calculate entropy of keypoints
func calculateKeypointEntropy(keypoints []gocv.KeyPoint, imageSize []int) float64 {
	gridSize := 10
	hist := make([]float64, gridSize*gridSize)
	for _, kp := range keypoints {
		x, y := int(kp.X*float64(gridSize)/float64(imageSize[1])), int(kp.Y*float64(gridSize)/float64(imageSize[0]))
		if x >= 0 && x < gridSize && y >= 0 && y < gridSize {
			hist[y*gridSize+x]++
		}
	}
	keypointEntropy := stat.Entropy(hist)
	return keypointEntropy
}

// Function to calculate grid-based uniformity score
func calculateGridUniformity(keypoints []gocv.KeyPoint, imageSize []int) float64 {
	height, width := imageSize[0], imageSize[1]
	gridSize := 10
	grid := make([]float64, gridSize*gridSize)
	for _, kp := range keypoints {
		x, y := int(kp.X*float64(gridSize)/float64(width)), int(kp.Y*float64(gridSize)/float64(height))
		if x >= 0 && x < gridSize && y >= 0 && y < gridSize {
			grid[y*gridSize+x]++
		}
	}

	gridMean := stat.Mean(grid, nil)
	gridStd := stat.StdDev(grid, nil)

	if gridMean == 0 {
		return 0 // Avoid division by zero
	}

	uniformityScore := 1 - (gridStd / gridMean)
	return uniformityScore
}

// Function to calculate the overall distribution score
func calculateDistributionScore(image gocv.Mat) float64 {
	// Detect and compute descriptors
	detector := gocv.NewSIFT()
	defer detector.Close()
	keypoints, descriptors := detectAndCompute(image, detector)

	// Image size
	imageSize := []int{image.Rows(), image.Cols()}

	// Calculate individual distribution metrics
	spatialDistribution := calculateSpatialDistribution(keypoints, imageSize)
	uniqueness := calculateUniquenessScore(descriptors)
	coverage := calculateCoverage(keypoints, imageSize)
	keypointEntropy := calculateKeypointEntropy(keypoints, imageSize)
	gridUniformity := calculateGridUniformity(keypoints, imageSize)

	// Normalize scores (example normalization factors)
	spatialDistributionNormalized := math.Min(spatialDistribution/10.0, 1.0)
	uniquenessNormalized := math.Min(uniqueness/100.0, 1.0)
	coverageNormalized := coverage
	keypointEntropyNormalized := math.Min(keypointEntropy/10.0, 1.0)
	gridUniformityNormalized := gridUniformity

	// Combine scores with weights
	weights := map[string]float64{
		"spatial_distribution": 0.2,
		"uniqueness":           0.2,
		"coverage":             0.2,
		"keypoint_entropy":     0.2,
		"grid_uniformity":      0.2,
	}

	combinedScore := weights["spatial_distribution"]*spatialDistributionNormalized +
		weights["uniqueness"]*uniquenessNormalized +
		weights["coverage"]*coverageNormalized +
		weights["keypoint_entropy"]*keypointEntropyNormalized +
		weights["grid_uniformity"]*gridUniformityNormalized

	finalDistributionScore := combinedScore * 100 // Scale to 0-100

	return finalDistributionScore
}

func main() {
	files := []string{
		"dataset/test_01.png",
		"dataset/test_02.png",
		"dataset/test_03.png",
		"dataset/test_04.png",
		"dataset/test_05.png",
		"dataset/test_09.png",
		"dataset/test_10.png",
		"dataset/test_11.png",
		"dataset/test_12.png",
		"dataset/test_13.png",
		"dataset/test_14.png",
		"dataset/test_15.png",
		"dataset/test_16.png",
		"dataset/test_17.png",
		"dataset/test_18.png",
		"dataset/test_19.png",
		"dataset/test_20.png",
		"dataset/test_21.png",
		"dataset/test_22.png",
		"dataset/test_23.png",
		"dataset/test_24.png",
		"dataset/test_25.png",
		"dataset/test_26.png",
		"dataset/test_27.png",
		"dataset/test_29.png",
		"dataset/test_30.png",
		"dataset/test_31.png",
		"dataset/test_32.png",
		"dataset/test_33.png",
		"dataset/test_34.png",
		"dataset/test_35.png",
		"dataset/test_36.png",
		"dataset/test_37.png",
		"dataset/test_38.png",
		"dataset/test_39.png",
		"dataset/test_40.png",
		"dataset/test_41.png",
		"dataset/test_42.png",
		"dataset/test_43.png",
		"dataset/test_44.png",
		"dataset/test_45.png",
		"dataset/test_46.png",
		"dataset/test_47.png",
		"dataset/test_49.png",
		"dataset/test_50.png",
		"dataset/test_51.png",
		"dataset/test_52.png",
		"dataset/test_53.png",
		"dataset/test_54.png",
		"dataset/test_55.png",
		"dataset/test_56.png",
		"dataset/test_57.png",
		"dataset/test_58.png",
	}

	for _, file := range files {
		fmt.Println("Processing:", file)
		image := gocv.IMRead(file, gocv.IMReadColor)
		if image.Empty() {
			log.Printf("Error: Unable to read %s", file)
			continue
		}
		defer image.Close()

		distributionScore := calculateDistributionScore(image)
		fmt.Printf("Image Distribution Score: %.2f\n", distributionScore)
	}
}

File: distribution.py
import cv2
import numpy as np
from sklearn.metrics import pairwise_distances
from scipy.stats import entropy


# Function to detect keypoints and compute descriptors
def detect_and_compute(image, detector):
    keypoints = detector.detect(image, None)
    keypoints, descriptors = detector.compute(image, keypoints)
    return keypoints, descriptors


# Function to calculate spatial distribution score
def calculate_spatial_distribution(keypoints, image_size):
    height, width = image_size
    x_coords = [kp.pt[0] for kp in keypoints]
    y_coords = [kp.pt[1] for kp in keypoints]
    x_hist, _ = np.histogram(x_coords, bins=10, range=(0, width))
    y_hist, _ = np.histogram(y_coords, bins=10, range=(0, height))
    spatial_distribution_score = entropy(x_hist) + entropy(y_hist)
    return spatial_distribution_score


# Function to calculate descriptor uniqueness score
def calculate_uniqueness_score(descriptors):
    distances = pairwise_distances(descriptors, metric="euclidean")
    min_distances = np.min(distances + np.eye(len(distances)) * distances.max(), axis=1)
    uniqueness_score = np.mean(min_distances)
    return uniqueness_score


# Function to calculate coverage score
def calculate_coverage(keypoints, image_size):
    height, width = image_size
    grid_size = 10
    grid = np.zeros((grid_size, grid_size))
    for kp in keypoints:
        x, y = kp.pt
        grid[int(y * grid_size / height), int(x * grid_size / width)] += 1
    coverage_score = np.count_nonzero(grid) / (grid_size * grid_size)
    return coverage_score


# Function to calculate entropy of keypoints
def calculate_keypoint_entropy(keypoints, image_size):
    height, width = image_size
    x_coords = [kp.pt[0] for kp in keypoints]
    y_coords = [kp.pt[1] for kp in keypoints]
    positions = np.vstack((x_coords, y_coords)).T
    hist, _, _ = np.histogram2d(
        positions[:, 0], positions[:, 1], bins=10, range=[[0, width], [0, height]]
    )
    keypoint_entropy = entropy(hist.flatten())
    return keypoint_entropy


# Function to calculate grid-based uniformity score
def calculate_grid_uniformity(keypoints, image_size):
    height, width = image_size
    grid_size = 10
    grid = np.zeros((grid_size, grid_size))
    for kp in keypoints:
        x, y = kp.pt
        grid[int(y * grid_size / height), int(x * grid_size / width)] += 1
    grid_mean = np.mean(grid)
    grid_std = np.std(grid)
    uniformity_score = 1 - (grid_std / grid_mean)
    return uniformity_score


# Function to calculate the overall distribution score
def calculate_distribution_score(image):
    # Detect and compute descriptors
    detector = cv2.SIFT_create()
    keypoints, descriptors = detect_and_compute(image, detector)

    # Image size
    image_size = image.shape[:2]

    # Calculate individual distribution metrics
    spatial_distribution = calculate_spatial_distribution(keypoints, image_size)
    uniqueness = calculate_uniqueness_score(descriptors)
    coverage = calculate_coverage(keypoints, image_size)
    keypoint_entropy = calculate_keypoint_entropy(keypoints, image_size)
    grid_uniformity = calculate_grid_uniformity(keypoints, image_size)

    # Normalize scores (example normalization factors)
    spatial_distribution_normalized = min(
        spatial_distribution / 10.0, 1.0
    )  # Example normalization factor
    uniqueness_normalized = min(uniqueness / 100.0, 1.0)  # Example normalization factor
    coverage_normalized = coverage  # Coverage is already between 0 and 1
    keypoint_entropy_normalized = min(
        keypoint_entropy / 10.0, 1.0
    )  # Example normalization factor
    grid_uniformity_normalized = (
        grid_uniformity  # Uniformity is already between 0 and 1
    )

    # Print
    # print(f"Spatial distribution score is : {spatial_distribution}")
    # print(f"Uniqueness normalized score is : {uniqueness_normalized}")
    # print(f"Coverage Normalized score is : {coverage_normalized}")
    # print(f"Keypoint entropy score is : {keypoint_entropy_normalized}")
    # print(f"Grid uniformity score is: {grid_uniformity_normalized}")
    # Combine scores with weights
    weights = {
        "spatial_distribution": 0.2,
        "uniqueness": 0.2,
        "coverage": 0.2,
        "keypoint_entropy": 0.2,
        "grid_uniformity": 0.2,
    }

    combined_score = (
        weights["spatial_distribution"] * spatial_distribution_normalized
        + weights["uniqueness"] * uniqueness_normalized
        + weights["coverage"] * coverage_normalized
        + weights["keypoint_entropy"] * keypoint_entropy_normalized
        + weights["grid_uniformity"] * grid_uniformity_normalized
    )

    # print(f"spatial distribution: {spatial_distribution_normalized}")
    # print(f"uniqueness: {uniqueness_normalized}")
    # print(f"coverage distribution: {coverage_normalized}")
    # print(f"grid uniformity: {grid_uniformity_normalized}")
    # print(f"keypoint entropy: {keypoint_entropy_normalized}")

    final_distribution_score = combined_score * 100  # Scale to 0-100

    return final_distribution_score


# # Load image
# files = [
#     "Images/test_01.jpeg",
#     "Images/test_02.jpeg",
#     "Images/test_03.jpeg",
#     "Images/test_04.jpeg",
#     "Images/test_05.jpeg",
#     "Images/test_06.jpeg",
#     "Images/test_07.jpg",
#     "Images/test_08.jpg",
#     "Images/test_09.png",
#     "Images/test_10.jpg",
#     "Images/test_11.png",
#     "Images/test_12.png",
#     "Images/test_13.png",
#     "Images/test_14.jpg",
#     "Images/test_15.png",
# ]
# for file in files:
#     print("Processing:", file)
#     image = cv2.imread(file)
#     if image is not None:
#         distribution_score = calculate_distribution_score(image)
#         print(f"Image Distribution Score: {distribution_score:.2f}")
#     else:
#         print(f"Error: Unable to read {file}")
